README.md
## ExpressJS API - Custom Implementation

: This project was intially developed to customized and ease the usage of configuring express framework.

---

## Installation

```bash
npm install expressjs-app
```

## Usage

```javascript
const { ExpressApi, LogOption } = require("expressjs-app")

const app = (router, ...args) => {
    router.get('/', (req, res) => {
        res.json({
            status: 200,
            message: "Hello, Shits!"
        })
    })

    return router
}

const headers = (req, res, next) => {
    res.set('Access-Control-Allow-Methods', 'GET,POST');
    next()
}

const unauthorize = (err, req, res, next) => {
    if (err.name === 'AuthenticationError') {
        // Send a customized response for authentication failure
        return res.status(err.status).json({ 
            status: err.status,
            message: 'Invalid Authorization header supplied.\nPlease contact support or check your credentials.'
        })
    }

    next()
}

const apiOptions = { 
    enableCors: true, 
    // lastRouteHandler: invalid, // Method for handling 404 not found routes
    // docsModule: swaggerDoc, // Swagger Docs
    enableLog: true, // Log option can also accept LogOption { path: '/logs' }
    // jwtStrategy: jwtStrategy // JWT Strategy, this package uses passport-jwt for authentication, if this provided as option make sure you enable `registerAuthenticationRoute`
}

const apiInstance = new ExpressApi(apiOptions)
apiInstance.registerRoutesGroup(
        "/api",
        {
            method: app, args: undefined, name: 'app'
        }
    )
    .registerMiddlewares(headers, unauthorize)
    // Start the API/App
    .start(async () => {
        console.log('Server Started')
    })

```

---

## Configuration

```env
NODE_ENV=staging
API_BASE_PATH=/
SECRET_KEY=pass123!
PORT=3000

# OPTIONAL
LOG_FORMAT=combined # default 'combined',
LOG_SIZE='10M' # default '10M',
LOG_INTERVAL='1d' # default '1d',
LOG_MAX_FILES=10 # default 10
```


CHANGELOG.md
### 1.3.8
- Fix the routing for single route registrations

### 1.4.1
- Added swagger docs config module

src/expressapi.js
const express = require('express');
const passport = require('passport');
const cors = require('cors')
const helmet = require('helmet')
const morgan = require('morgan')
const { getFileStream, config, isAsync, LogOption } = require('./util');
const { Swagger } = require('./components/swagger');

/**
 * 
 * ## Implementation expressJS for custom use
 * @author Nino Casupananq
 * 
 * @param enableCors If cors will be enabled
 * @param lastRouteHandler last route handler registration
 * @param docsModule module for the documentation - only available for Swagger docs
 * @param enableLog flag to enable or disable log. Default is `false`
 */
module.exports = class ExpressApi {
    constructor({ enableCors, lastRouteHandler, docsModule, enableLog, jwtStrategy } = { enableCors: true, lastRouteHandler: (req, res) => {}, docsModule: (server) => {}, enableLog: LogOption(), jwtStrategy: undefined }) {
        this.server = express()
        this.port = process.env.PORT || 3000
        this.router = express.Router()
        this.groupRouters = {}
        this.passport = passport
        this.lastRouteHandler = undefined
        this.jwtStrategy = undefined

        if (enableCors) this.server.use(cors())
        if (typeof lastRouteHandler === 'function') this.lastRouteHandler = lastRouteHandler
        if (typeof docsModule === 'function') {
            docsModule(this.server)
            console.info(`[${process.env.NODE_ENV}][ExpressApi] Registered docs`)
        }

        if (typeof docsModule === 'object') {
            const swagger = new Swagger(docsModule)
            swagger.serveDocs(this.server)
            console.info(`[${process.env.NODE_ENV}][ExpressApi] Registered docs`)
        }

        if (jwtStrategy != undefined) {
            this.jwtStrategy = jwtStrategy
        }

        // Helment adds required security headers in the middleware
        this.server.use(helmet())

        // The express.json() function is a built-in middleware function in Express. It parses incoming requests with JSON payloads and is based on body-parser.
        this.server.use(express.json())

        if (enableLog == true) {
            // Register logger
            this.createLogFor('access')
            this.createLogFor('error')
            console.info(`[${process.env.NODE_ENV}][ExpressApi] Logging enabled.`)
        } else if (enableLog instanceof LogOption.constructor) {
            this.createLogFor('access', enableLog.path)
            this.createLogFor('error', enableLog.path)
            console.info(`[${process.env.NODE_ENV}][ExpressApi] Logging enabled.`)
        }
    }


    createLogFor(type = 'access', path = undefined) {
        // const logFile = createWriteStream(`./${type}.log`, { flags: 'a' });

        const options = { stream: getFileStream(type, path) }

        options.skip = function (req, res) {
            return res.statusCode > 301
        }

        if (type.includes('err')) options.skip = function (req, res) { return res.statusCode < 301 }
        this.server.use(morgan(config.LOG_FORMAT, options))
    }


    /**
     * ### Register routes
     * @param {*} route pass the route method that accepts `Router` as argument
     * @param  {*} args Additional arguments for the route method as Object
     * @param {*} name Name a route
     * @returns ExpressApi
     * 
     * @example
     * ```javascript
     * 
     * const route = (router) => {
     *      router.get('/', (req, res) => {
     *          // process
     *      })
     *      
     *      // return router is required.
     *      return router
     * }
     * 
     * const api = new ExpressApi();
     * api.registerRoutes(route).start()
     * ```
     */
    registerRoutes(...routes) {
        // route = () => {}, args = {}, name
        routes.map(route => {
            this.router = route.method(this.router, route.args)
            console.log(`[${process.env.NODE_ENV}][ExpressApi] ${route.name || 'RouteName'} - route is registered.`)
        })

        return this
    }

    /**
     * Route Groups is best for organizing routes and handlers
     * @param {*} baseUri 
     * @param  {...any} routes 
     * @returns 
     * 
     * @example
     * ```javascript
     * const route1 = (router) => {
     *      router.get('/', (req, res) => {
     *          // process
     *      })
     *      
     *      // return router is required.
     *      return router
     * }
     * 
     * const route2 = (router) => {
     *      router.get('/', (req, res) => {
     *          // process
     *      })
     *      
     *      // return router is required.
     *      return router
     * }
     * 
     * const api = new ExpressApi();
     * api.registerRouteGroup('/v1/api', route1, route2).start()
     * ```
     */
    registerRoutesGroup(baseUri = '/api', ...routes) {
        let router = express.Router()
        routes.forEach(route => {
            router = route.method(router, route.args)
            console.log(`[${process.env.NODE_ENV}][ExpressApi] ${route.name || 'RouteName'} - route is registered.`)
        })

        console.info(`[${process.env.NODE_ENV}][ExpressApi] Configured baseUri ${baseUri}`)

        // Add router
        this.groupRouters[baseUri] = router

        return this
    }


    /**
     * ### Register authentication - Call this method if you want to add the authentication otherwise dont use.
     * @param {*} authRoute pass the auth route that accepts `express()` instance. To exclude the auth route.
     * @param {*} protectExpr set the protected path - wildcard
     * @returns ExpressApi
     * 
     * @example
     * 
     * ```javascript
     * const authRoute = (server) => {
     *      server.post('/authenticate', (req, res) => {
     *          // authenticate user
     *      })
     * }
     * 
     * const api = new ExpressApi();
     * api.registerAuthentication(authRoute, '/api/*').start()
     * ```
     */
    registerAuthenticationRoute(baseUri = '/api', authRoute = () => { }, protectExpr = "*") {
        if (this.jwtStrategy == undefined) throw Error("jwtStrategy is not properly defined.")
        // this.server.use(express.json());
        this.server.use(this.passport.initialize());
        this.passport = this.passport.use(this.jwtStrategy)

        this.server = authRoute(baseUri, this.server)

        this.server.all((protectExpr.includes("*") ? protectExpr : `${protectExpr}**`), this.passport.authenticate('jwt', { session: false, failWithError: true }))
        
        console.log(`[${process.env.NODE_ENV}][ExpressApi] Authentication is configured to this path: ${protectExpr}`)

        return this
    }


    /**
     * ### Register a middleware - `app.use`
     * @param {*} cb function to register
     * @returns ExpressApi
     * 
     * @example
     * 
     * ```javascript
     * const api = new ExpressApi();
     * api.registerMiddleware((req, res, next) => {
     *      // process
     * }).start()
     * ```
     */
    registerMiddlewares(...args) {
        args.map(fn => this.server.use(fn))
        return this
    }

    /**
     * ### Start the server instance
     * @param {*} callback set the callback for `listen`
     * @param {*} port port of the server
     */
    start(callback = async () => { }, port = 0) {
        if (port != 0) this.port = port

        if (Object.keys(this.groupRouters).length) {
            for (let uri in this.groupRouters) {
                this.server.use(uri, this.groupRouters[uri])
            }
        } else {
            this.server.use(this.router)
        }

        if (this.lastRouteHandler !== undefined && typeof this.lastRouteHandler === 'function') this.server.use(this.lastRouteHandler)

        this.server.listen(this.port, async () => {
            console.log(`[${process.env.NODE_ENV}][ExpressApi] API Started @ port ${this.port}`)
            
            if (isAsync(callback)) {
                await callback()
            } else {
                callback()
            }
        })
    }
}

src/components/uriBuilder.js

/**
 * 
 * UriBuilder will be use to format uri building instead of appending long URIs before
 * sending to Axios.
 * 
 * @param basePath String
 */
module.exports.UriBuilder = class UriBuilder {
    constructor(basePath = "") {
        this.uri = ""
        this.uri += (basePath.includes("/") && this.uri != "" ? basePath.substring(1) : basePath) + "?"
        this.params = new URLSearchParams()
    }

    addQuery(key, value) {
        this.params.append(key, value)
        this.uri += `${(this.isNoQueryParam())}${key}=${value}`
        return this
    }

    addRaw(raw) {
        this.uri += `${(this.isNoQueryParam())}${raw}`
        return this
    }

    isNoQueryParam() {
        return (this.uri.substring(this.uri.length - 1) == "?" ? "" : "&")
    }

    getUri() {
        return this.uri
    }

    getParams() {
        return this.params
    }
}


src/util.js
const { existsSync, mkdirSync } = require('fs')
const { join }  = require('path')
const rfs = require('rotating-file-stream')
const logFolder = process.env.LOG_FOLDER || join(process.cwd(), 'logs')

const config = {
    LOG_FORMAT: process.env.LOG_FORMAT || 'combined',
    LOG_SIZE: process.env.LOG_SIZE ||'10M',
    LOG_INTERVAL: process.env.LOG_INTERVAL ||'1d',
    LOG_MAX_FILES: parseInt(process.env.LOG_MAX_FILES ||10),
    REQUEST_CACHING_TTL: parseInt(process.env.REQUEST_CACHING || 30)
}


const getFileStream = (type, path) => {
    const logPath = path !== undefined ? path : logFolder
    const createFileName = (time, index) => {
        if (!time) {
            return `${type}-current.log`
        }

        let filename = time.toISOString().slice(0, 10);
        if (index > 1) {
            filename += `.${index}`;
        }

        return join(logPath, `${type}-${filename}.log.gz`);
    };

    return rfs.createStream(createFileName, {
        interval: config.LOG_INTERVAL,
        maxSize: config.LOG_SIZE,
        maxFiles: config.LOG_MAX_FILES,
        initialRotation: true,
        path: logPath,
        compress: 'gzip'
    })
}

/**
 * 
 * HTTP Status Codes
 */
const HttpStatus = {
    CONTINUE: 100,
    SWITCHING_PROTOCOLS: 101,
    PROCESSING: 102,
    EARLY_HINTS: 103,
    OK: 200,
    CREATED: 201,
    ACCEPTED: 202,
    NON_AUTHORITATIVE_INFORMATION: 203,
    NO_CONTENT: 204,
    RESET_CONTENT: 205,
    PARTIAL_CONTENT: 206,
    MULTI_STATUS: 207,
    ALREADY_REPORTED: 208,
    IM_USED: 226,
    MULTIPLE_CHOICES: 300,
    MOVED_PERMANENTLY: 301,
    FOUND: 302,
    SEE_OTHER: 303,
    NOT_MODIFIED: 304,
    USE_PROXY: 305,
    TEMPORARY_REDIRECT: 307,
    PERMANENT_REDIRECT: 308,
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    PAYMENT_REQUIRED: 402,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    METHOD_NOT_ALLOWED: 405,
    NOT_ACCEPTABLE: 406,
    PROXY_AUTHENTICATION_REQUIRED: 407,
    REQUEST_TIMEOUT: 408,
    CONFLICT: 409,
    GONE: 410,
    LENGTH_REQUIRED: 411,
    PRECONDITION_FAILED: 412,
    PAYLOAD_TOO_LARGE: 413,
    URI_TOO_LONG: 414,
    UNSUPPORTED_MEDIA_TYPE: 415,
    RANGE_NOT_SATISFIABLE: 416,
    EXPECTATION_FAILED: 417,
    IM_A_TEAPOT: 418,
    MISDIRECTED_REQUEST: 421,
    UNPROCESSABLE_ENTITY: 422,
    LOCKED: 423,
    FAILED_DEPENDENCY: 424,
    TOO_EARLY: 425,
    UPGRADE_REQUIRED: 426,
    PRECONDITION_REQUIRED: 428,
    TOO_MANY_REQUESTS: 429,
    REQUEST_HEADER_FIELDS_TOO_LARGE: 431,
    UNAVAILABLE_FOR_LEGAL_REASONS: 451,
    INTERNAL_SERVER_ERROR: 500,
    NOT_IMPLEMENTED: 501,
    BAD_GATEWAY: 502,
    SERVICE_UNAVAILABLE: 503,
    GATEWAY_TIMEOUT: 504,
    HTTP_VERSION_NOT_SUPPORTED: 505,
    VARIANT_ALSO_NEGOTIATES: 506,
    INSUFFICIENT_STORAGE: 507,
    LOOP_DETECTED: 508,
    NOT_EXTENDED: 510,
    NETWORK_AUTHENTICATION_REQUIRED: 511
}

const HttpStatusMessage = {
    100: "CONTINUE",
    101: "SWITCHING_PROTOCOLS",
    102: "PROCESSING",
    103: "EARLY_HINTS",
    200: "OK",
    201: "CREATED",
    202: "ACCEPTED",
    203: "NON_AUTHORITATIVE_INFORMATION",
    204: "NO_CONTENT",
    205: "RESET_CONTENT",
    206: "PARTIAL_CONTENT",
    207: "MULTI_STATUS",
    208: "ALREADY_REPORTED",
    226: "IM_USED",
    300: "MULTIPLE_CHOICES",
    301: "MOVED_PERMANENTLY",
    302: "FOUND",
    303: "SEE_OTHER",
    304: "NOT_MODIFIED",
    305: "USE_PROXY",
    307: "TEMPORARY_REDIRECT",
    308: "PERMANENT_REDIRECT",
    400: "BAD_REQUEST",
    401: "UNAUTHORIZED",
    402: "PAYMENT_REQUIRED",
    403: "FORBIDDEN",
    404: "NOT_FOUND",
    405: "METHOD_NOT_ALLOWED",
    406: "NOT_ACCEPTABLE",
    407: "PROXY_AUTHENTICATION_REQUIRED",
    408: "REQUEST_TIMEOUT",
    409: "CONFLICT",
    410: "GONE",
    411: "LENGTH_REQUIRED",
    412: "PRECONDITION_FAILED",
    413: "PAYLOAD_TOO_LARGE",
    414: "URI_TOO_LONG",
    415: "UNSUPPORTED_MEDIA_TYPE",
    416: "RANGE_NOT_SATISFIABLE",
    417: "EXPECTATION_FAILED",
    418: "IM_A_TEAPOT",
    421: "MISDIRECTED_REQUEST",
    422: "UNPROCESSABLE_ENTITY",
    423: "LOCKED",
    424: "FAILED_DEPENDENCY",
    425: "TOO_EARLY",
    426: "UPGRADE_REQUIRED",
    428: "PRECONDITION_REQUIRED",
    429: "TOO_MANY_REQUESTS",
    431: "REQUEST_HEADER_FIELDS_TOO_LARGE",
    451: "UNAVAILABLE_FOR_LEGAL_REASONS",
    500: "INTERNAL_SERVER_ERROR",
    501: "NOT_IMPLEMENTED",
    502: "BAD_GATEWAY",
    503: "SERVICE_UNAVAILABLE",
    504: "GATEWAY_TIMEOUT",
    505: "HTTP_VERSION_NOT_SUPPORTED",
    506: "VARIANT_ALSO_NEGOTIATES",
    507: "INSUFFICIENT_STORAGE",
    508: "LOOP_DETECTED",
    510: "NOT_EXTENDED",
    511: "NETWORK_AUTHENTICATION_REQUIRED"
}

module.exports.isAsync = fn => {
    const AsyncFunc = (async () => {}).constructor;
    return fn instanceof AsyncFunc
}

module.exports.LogOption = {
    path: logFolder
}

module.exports.getFileStream = getFileStream
module.exports.config = config
module.exports.HttpStatus = HttpStatus
module.exports.HttpStatusMessage = HttpStatusMessage